python/pyarrow/_compute.pyx

```diff
@@ -2251,3 +2325,158 @@ cdef CExpression _bind(Expression filter, Schema schema) except *:
 
     return GetResultValue(filter.unwrap().Bind(
         deref(pyarrow_unwrap_schema(schema).get())))
+
+
+cdef CFunctionDoc _make_function_doc(dict func_doc) except *:
+    """
+    Helper function to generate the FunctionDoc
+    """
+    cdef:
+        CFunctionDoc f_doc
+        vector[c_string] c_arg_names
+        c_bool c_options_required
+
+    validate_expr = "summary" in func_doc.keys(
+    ) and "description" in func_doc.keys() and "arg_names" in func_doc.keys()
+    if not validate_expr:
+        raise ValueError(
+            "function doc dictionary must contain, summary, arg_names and a description")
+    f_doc.summary = tobytes(func_doc["summary"])
+    f_doc.description = tobytes(func_doc["description"])
+    for arg_name in func_doc["arg_names"]:
+        c_arg_names.push_back(tobytes(arg_name))
+    f_doc.arg_names = c_arg_names
+    # UDFOptions integration:
+    # TODO: https://issues.apache.org/jira/browse/ARROW-16041
+    f_doc.options_class = tobytes("None")
+    c_options_required = False
+    f_doc.options_required = c_options_required
+    return f_doc
+
+
+def register_scalar_function(func_name, num_args, function_doc, in_types,
+                             out_type, function):
+    """
+    Register a user-defined-function.
+
+    Parameters
+    ----------
+
+    func_name : str
+        Name of the function. This name must be globally unique. 
+    num_args : int
+       Number of arguments in the function.
+       When defining a function with variable arguments, 
+       the num_args represents the minimum number of arguments
+       required. 
+    function_doc : dict
+        A dictionary object with keys "summary" (str),
+        "description" (str), and "arg_names" (list of str).
+    in_types : List[InputType]
+        List of InputType objects which defines the input 
+        types for the function. When defining a list of InputType
+        for a varargs function, the list only needs to contain the
+        number of elements equal to the num_args (which is the miniumu
+        required arguments).
+    out_type : DataType
+        Output type of the function.
+    function : callable
+        User-defined-function
+        function includes arguments equal to the number
+        of input_types defined. The return type of the 
+        function is of the type defined as output_type. 
+        The output should be an Array, Scalar, ChunkedArray,
+        Table, or RecordBatch based on the out_type.
```

Comment:
I think this was noted already, but do we expose ChunkedArray, Table, or RecordBatch output types?